### ISSUE TITLE
Closing prompt with ESC or letting it time out is treated as a password entry attempt

### ISSUE THREAD

**Issue by Diomendius** [google/xsecurelock#114]
Since a recent system update, the `xsecurelock` password prompt will, upon closing, show "Processing..." and remain unresponsive for several seconds, as if an incorrect password was entered. This is not equivalent to pressing return, however; entering the correct password and either pressing `ESC` or waiting for the prompt to time out does not unlock the screen. It does, however, count as an incorrect login attempt for the purposes of `faillock`, as opening and closing the prompt three times with `ESC` (with default `faillock` configuration) causes the account to be locked for ten minutes.

I don't know specifically what system change caused this, but my best guess is something related to PAM, as `xsecurelock` has not been updated since January.

I am using Arch Linux with packages up to date as of 2020-09-28, `xsecurelock` version 1.7.0 and PAM version 1.4.0.

**Comment by dadevel** [comment]
I noticed the same bug today. I am using Arch Linux as well. With `xsecurelock` v1.7.0 and `pam` v1.4.0.

**Comment by lkrms** [comment]
I’ve noticed this recently, too. And I haven’t been able to replicate it yet, but I suspect a mouse click was also interpreted as an attempt. 

This has significant repercussions when using systems that default to 10minute lockouts after 3 failed attempts (eg Arch).

**Comment by Sonico98** [comment]
Yes, simply waking up the screen by moving the cursor results in incorrect login attempts.
I'm using Arch Linux too, `xsecurelock 1.7.0` from the repos and `pam 1.5.1`

**Comment by divVerent** [comment]
This might not be avoidable other than by _killing_ the authentication process when escape is hit.

Currently, on escape, we send `PAM_CONV_ERR` to PAM when the user presses escape. I'd assume that faillog shouldn't count this as a failure as it is, by the error code, not the password that failed but the conversation (i.e. talking to the user).

However, looking at the source code of pam_tally, the faillog implementation is not based on the status code at all. If the PAM module fails for any reason, it increments the failure counter.

Hm... the only workaround I could offer is "uncleanly" exiting the PAM session not by returning conversation error but by calling exit(1) right away. This is dirty and some PAM modules may not properly release resources when doing this, so I am rather reluctant to doing that.

Know any other software that does this?

**Comment by Diomendius** [comment]
I don't think PAM always treated this as an incorrect attempt. There was definitely an observable difference between an incorrect password attempt and `ESC`/timeout at some point in the recent past, and I'm fairly certain this happened with the same version of `xsecurelock`. It might be worth bisecting PAM, or looking at the blame logs if you have a more specific idea of what could have changed.

**Comment by divVerent** [comment]
The PAM conversation has started at that point but not ended; we could of
course just kill the auth process on esc (then it would not show as a bad
attempt) but this sounds like it generally allows circumventing logging of
bad attempts.
